### PR TITLE
Big test

### PR DESCRIPTION
**Pull request by Quaqqer**
Added a python script that generates sylt code for performance testing. I think it's working quite well and it can bring the sylt compiler to its knees if you bump up block_assignments, nested_expression_limit and nested_function_limit. The block assignments is currently not correctly calculated, since I don't fully understand how the sylt compiler compiles nested expressions (or most likely I'm just too lazy to count).

It currently also writes to both stdout and to a file called big_test.sy in cwd, probably want to change that.

Also the nested_expression_limit as well as the nested_expression_limit may be calculated a bit wrong.

Related to issue #343.

### CODE REVIEW

**Comment by Quaqqer** [comment]
Here is some example output

```start :: fn  ->
  
end
v1 :: 67
v2 :: fn a ->
  v3 :: 77
  v4 :: 16
  v5 :: (a + 44)
  v6 := ((55 * 83) + 27)
  v7 := 19
  v8 :: 19
  v9 := (55 * v5)
  v10 := v8
  v11 :: a
  v12 :: 92
  v13 := (47 * (33 * 5))
  v14 :: (36 - (v3 * 93))
  v15 := ((20 + v3) + (98 * a))
  v16 := ((21 * 98) + (45 - 12))
  v17 :: 60
  v18 := 25
  v19 := 69
  v20 := (38 - v4)
  v21 := 78
  v22 := (2 - (61 + 1))
  v23 := (37 - 5)
  v3
end
v25 := fn a ->
  v26 := v1
  v27 := (v2(88) - v2(v1))
  v28 :: v2(a)
  v29 :: 60
  v30 :: 93
  v31 :: v29
  v32 :: v2((9 * v30))
  v33 :: ((v29 * 82) + (40 - 65))
  v34 :: v1
  v35 := ((46 * a) + v31)
  v36 := 0
  v37 :: (v33 * v2(46))
  v38 :: (v2(37) - (35 * 41))
  v39 :: 85
  v40 :: 71
  v41 :: v39
  v42 :: (a - v2(v38))
  v43 := v2(v2(v39))
  v44 :: v2((v34 - v40))
  v45 := v30
  v46 := v2(v2(42))
  v47 :: v45
  v48 :: v2(v2(v44))
  v49 :: ((67 * v45) * v31)
  v50 := (v41 * (20 * 66))
  v40
end
v52 := fn a ->
  v53 :: 46
  v54 :: 64
  v55 :: v25(62)
  v56 := v25(v25(6))
  v57 := v2(97)
  v58 :: v53
  v59 := ((38 + 1) * v25(70))
  v60 := v2(v25(84))
  v61 := v25(25)
  v62 :: 42
  v63 :: v2((v55 + 55))
  v64 := 74
  v65 := 77
  v66 := 88
  v67 :: 67
  v68 := (v25(v59) + (v67 * 77))
  v69 := v59
  v70 := 82
  v71 :: 30
  v72 :: 95
  v73 := v25(v2(58))
  v74 := v2(v59)
  v75 :: (v25(53) - v2(80))
  v76 :: (v2(3) * v25(36))
  v77 :: (v25(v54) * (v54 + 19))
  v78 :: 40
  v79 := v25(35)
  v80 :: v25(v25(v68))
  v81 :: (v2(11) * v59)
  v68
end
v83 := fn a ->
  v84 := v52((40 - a))
  v85 :: v52((38 + v84))
  v86 :: (v52(57) - v25(v1))
  v87 :: ((v84 - 91) + (34 + 14))
  v88 :: v87
  v89 := 52
  v90 := v87
  v91 :: ((a * v86) + v1)
  v92 :: (v89 - v25(80))
  v93 := v52(41)
  v94 :: (v25(v1) + v25(v87))
  v95 := v2(70)
  v96 :: v25(v25(38))
  v97 := v92
  v98 := v25((15 - 57))
  v99 :: v2(4)
  v100 := v2(v25(28))
  v101 :: ((19 + 78) - (63 + 35))
  v102 :: v25(v88)
  v103 :: 63
  v104 :: ((70 - 37) - (44 - 64))
  v105 :: (v2(73) + v25(40))
  v106 :: (v52(36) - 24)
  v95
end
v108 :: (42 * (96 - 10))
v109 :: (v2(63) + v83(38))
v110 := ((v1 + v108) + v2(v109))
v111 :: v83(58)
v112 := fn a ->
  v113 :: 19
  v114 := v25(v83(39))
  v115 :: (90 * (v114 + 43))
  v116 := ((v111 + 33) - v83(56))
  v117 :: (v113 + (17 * v113))
  v118 := (v114 * v25(62))
  v119 :: (a - 72)
  v120 :: v108
  v121 :: ((65 + 16) - v2(10))
  v122 := (89 - v52(16))
  v123 :: 19
  v124 := 62
  v125 :: ((17 - 31) * v25(66))
  v126 := v25((26 * v121))
  v127 := 45
  v128 :: v83((33 + v123))
  v129 := v25(v83(v125))
  v130 :: v52((64 - v121))
  v131 := v52(v52(82))
  v132 := v52(45)
  v133 :: v83(v2(78))
  v134 :: 67
  v135 :: (v83(27) - 21)
  v136 := 2
  v137 := 4
  v111
end
v139 :: 6
v140 :: fn a ->
  v141 :: ((5 - 52) * v25(70))
  v142 := v109
  v143 :: 87
  v144 := v109
  v145 :: ((29 - v143) - 98)
  v146 :: (v83(52) - v52(72))
  v147 :: v83((12 + v108))
  v148 :: v112(v52(68))
  v149 :: a
  v150 := (v1 - v146)
  v151 :: v52(87)
  v152 := 28
  v153 := v2(v25(3))
  v154 :: v139
  v155 :: v112(19)
  v156 := v52((36 - 73))
  v157 := 93
  v158 := 15
  v159 :: v2((74 + 63))
  v160 :: v83((v151 - v109))
  v161 := a
  v162 := v25(v52(v141))
  v163 := v25((14 + 62))
  v164 :: 93
  v165 :: (32 + (0 + 51))
  v166 := v2((v158 - v152))
  v167 :: v2(47)
  v168 := v83(v153)
  v169 := 26
  v170 := 38
  v145
end
v172 :: fn a ->
  v173 :: (v140(12) - (43 * v139))
  v174 := v140(8)
  v175 :: v83(74)
  v176 := 50
  v177 :: v83(83)
  v178 :: v25(v139)
  v179 :: ((82 * v177) * v52(v110))
  v180 :: (v25(v1) + v174)
  v181 := v2(v25(v108))
  v182 :: v139
  v183 := v175
  v184 :: ((0 * 40) * v25(85))
  v185 := 96
  v186 :: v112(0)
  v187 := 34
  v188 :: ((88 + 30) + v140(v177))
  v189 := v25(v25(27))
  v190 := (79 * 88)
  v191 :: 76
  v192 := v25((97 - 75))
  v193 :: ((69 - 60) + (78 - 11))
  v194 :: (55 - 1)
  v195 :: v179
  v196 :: (20 - v25(20))
  v111
end
v198 :: fn a ->
  v199 :: (17 - (7 - 98))
  v200 := ((a + 36) + (38 * 3))
  v201 := v1
  v202 := v172(v83(60))
  v203 :: a
  v204 :: ((v139 * a) * v202)
  v205 :: 89
  v206 :: ((v1 + v108) * (v1 + 26))
  v207 := (v112(94) - (5 * 38))
  v208 := v83(v83(30))
  v209 := ((v200 * 62) - (v108 * v203))
  v210 := v2(84)
  v211 :: v112((69 - 66))
  v212 :: 95
  v213 :: 71
  v214 := (63 - v172(71))
  v215 :: 10
  v216 := ((34 + 5) + v140(v213))
  v217 := ((15 * 65) - v208)
  v109
end
v219 := v25(v25(67))
v220 :: v52(v2(v109))
v221 := v52(v172(56))
v222 :: fn a ->
  v223 :: ((69 - 91) * 5)
  v224 :: v172(a)
  v225 :: v140((25 - 44))
  v226 :: (64 * (61 - v221))
  v227 := v25(v83(v139))
  v228 :: v140(v223)
  v229 := 28
  v230 :: (v112(89) + v112(70))
  v231 := v140(v198(90))
  v232 :: 16
  v233 := (49 * (9 + 20))
  v234 := ((31 - 27) - (94 - 38))
  v235 := 20
  v236 :: v232
  v237 :: v112(v25(v219))
  v238 := v83((66 + 50))
  v239 :: 75
  v240 := (v25(67) * (12 * 26))
  v241 := v172((v226 + v225))
  v242 := ((76 - 79) + v112(86))
  v243 :: (92 - v172(2))
  v244 :: 72
  v245 := 33
  v246 :: v241
  v221
end
v248 :: ((v109 - 15) + (22 - 31))
v249 :: ((v108 - 3) * (v139 - 92))
v250 := 14
v251 := fn a ->
  v252 := (v222(49) + (32 * 60))
  v253 := v198(13)
  v254 :: 55
  v255 :: v83(v222(v252))
  v256 := v172(73)
  v257 := v255
  v258 :: v140((v253 * v255))
  v259 :: v140(v25(12))
  v260 :: (v25(93) + (48 - 11))
  v261 :: v249
  v262 := ((v255 + v220) * (5 + v254))
  v263 := v2((28 + 9))
  v264 :: 13
  v265 := v222(v2(94))
  v266 :: 48
  v267 :: 75
  v268 := (v172(54) + (19 - 43))
  v269 :: v172(v140(37))
  v270 :: ((2 - v266) + (41 + 31))
  v271 :: (v52(70) - (v268 - v110))
  v272 :: (v140(93) + (34 + v220))
  v273 :: v52(v222(v272))
  v1
end
v275 :: v219
v276 := 41
v277 := fn a ->
  v278 := (v249 - v172(v108))
  v279 :: v140((v1 + v220))
  v280 :: v140(v2(v111))
  v281 :: v251((32 - 50))
  v282 :: 51
  v283 := v112((v139 + v280))
  v284 :: (v140(65) * v198(v280))
  v285 :: v248
  v286 := 97
  v287 :: (75 * 14)
  v288 := v248
  v289 := ((35 * 16) + v25(46))
  v290 :: 78
  v291 := 73
  v292 := (v172(95) * 84)
  v293 := ((23 - 3) * 79)
  v294 :: (70 + v198(90))
  v295 := (v25(54) + (87 - 73))
  v296 :: v172(v25(89))
  v297 :: v251((86 - v296))
  v298 := 39
  v299 :: (92 - v83(83))
  v300 := ((18 * v280) - (4 + 37))
  v285
end
v302 :: fn a ->
  v303 :: v52(v140(99))
  v304 := v277((v276 * v303))
  v305 :: v25((v276 + 70))
  v306 := v277(40)
  v307 := v251(88)
  v308 :: (v222(v219) + v277(83))
  v309 := v277((15 + 59))
  v310 :: ((40 * 12) - v222(38))
  v311 := ((65 + 36) - (81 - 91))
  v312 :: v250
  v313 :: v251((30 * 7))
  v314 :: (v172(v220) + v2(51))
  v315 :: v140((0 - 45))
  v316 :: v110
  v317 :: 78
  v318 := (51 * v251(v304))
  v319 := v172(v307)
  v320 :: 13
  v321 :: v140(v1)
  v322 :: v2((v221 + v310))
  v323 := (96 - v2(21))
  v324 := (v140(84) * 54)
  v325 :: (v83(4) - v25(v313))
  v326 :: (v172(65) * 75)
  v327 :: v317
  v328 := v318
  v329 := v315
  v139
end
v331 :: 47
v332 :: ((v219 * v220) - (v250 + 42))
v333 := fn a ->
  v334 := 64
  v335 :: (v198(63) * (v111 * 61))
  v336 := (v250 - (v108 * 9))
  v337 := ((v109 * 60) * 16)
  v338 := ((v275 - 43) - 49)
  v339 :: 27
  v340 := v302(3)
  v341 :: v1
  v342 :: (v222(74) - 42)
  v343 := v277(v112(32))
  v344 := 63
  v345 := 44
  v346 := (55 * (v341 - v220))
  v347 :: v277(88)
  v348 := 25
  v349 :: 31
  v350 :: v83(v251(v340))
  v351 :: v348
  v352 :: v219
  v353 :: v198(v139)
  v354 :: ((35 - 99) - v52(55))
  v355 := v25((84 - v111))
  v356 := 2
  v357 :: (v83(57) - 66)
  v358 := v172(v351)
  v248
end
v360 := v248
```

File: random_test.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+Future improvements:
+    - Guarantee longer dependency chains by not just randomly choosing
+      a function/value
+"""
+
+
+from random import choice, random, shuffle
+from typing import Union
+
+indented = lambda s: "\n".join(map(lambda ss: "  " + ss, s.splitlines())) + "\n"  # noqa
```
Comment (FredTheDino) [inline random_test.py]: If you want to add a newline to the end, you can put it at the end of the string concatination no?
```suggestion
indented = lambda s: map(lambda ss: f"  {ss}\n", s.splitlines()))
```
(I didn't try this)

**Review by FredTheDino — APPROVED** [review]
This looks pretty poggers! I wonder a bit what the size member function is for, but other than that this looks pretty great. Maybe it's time to generate some random programs and run them in the CI, and also to do some benchmarking for them. :o

To me it looks like the code generates a fully valid sylt-program, since all of it is numbers, so it should pass the typechecker.

I tried to run one of these programs and the parser pretty much died.

Earlier versions had a problem with being very slow to run, this latest commit looks to have fixed it. Great work!

**Comment by Quaqqer** [comment]
> I wonder a bit what the size member function is for, but other than that this looks pretty great.

Since lua has a hard limit of 200 local variables I thought it would be nice to guarantee that the randomly generated programs actually run with lua. The size is calculated according to how the sylt compiler compiles nested expressions.

For instance:
```
a :: (b + c) * d
```
Compiles to:
```lua
local intermediate = b + c
local a = intermediate * d
```

(Or something along those lines)

**Review by EMattfolk — APPROVED** [review]

**Pull request merged**